src/main/java/org/opensearch/knn/index/codec/derivedsource/NestedPerFieldDerivedVectorInjector.java

```diff
@@ -0,0 +1,276 @@
+     * @throws IOException if there is an issue reading from the formats
+     */
+    private int mapToDocId(Map<String, Object> doc, int offset, int parent) throws IOException {
+        // For all the fields, we look for the first doc that matches any of the fields.
+        int position = NO_MORE_DOCS;
+        for (String key : doc.keySet()) {
+            position = getLowestDocIdForField(ParentChildHelper.constructSiblingField(childFieldInfo.name, key), offset);
+            if (position < parent) {
+                break;
+            }
+        }
+
+        // Advancing past the parent means something went wrong
+        assert position < parent;
+        return position;
+    }
+
+    /**
+     * Get the lowest docId for a field that is greater than the offset.
+     *
+     * @param fieldToMatch field to find the lowest docId for
+     * @param offset       offset to start searching from
+     * @return lowest docId for the field that is greater than the offset. Returns {@link DocIdSetIterator#NO_MORE_DOCS} if doc cannot be found
+     * @throws IOException if there is an issue reading from the formats
+     */
+    private int getLowestDocIdForField(String fieldToMatch, int offset) throws IOException {
+        // This method implementation is inspired by the FieldExistsQuery in Lucene and the FieldNamesMapper in
+        // Opensearch. We first mimic the logic in the FieldExistsQuery in order to identify the docId of the nested
+        // doc. If that fails, we rely on
+        // References:
+        // 1. https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/FieldExistsQuery.java#L170-L218.
+        // 2.
+        // https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/mapper/FieldMapper.java#L316-L324
+        FieldInfo fieldInfo = segmentReadState.fieldInfos.fieldInfo(fieldToMatch);
+
+        if (fieldInfo == null) {
+            return NO_MORE_DOCS;
+        }
+
+        DocIdSetIterator iterator = null;
+        if (fieldInfo.hasNorms() && derivedSourceReaders.getNormsProducer() != null) { // the field indexes norms
+            iterator = derivedSourceReaders.getNormsProducer().getNorms(fieldInfo);
+        } else if (fieldInfo.getVectorDimension() != 0 && derivedSourceReaders.getKnnVectorsReader() != null) { // the field indexes vectors
+            switch (fieldInfo.getVectorEncoding()) {
+                case FLOAT32:
+                    iterator = derivedSourceReaders.getKnnVectorsReader().getFloatVectorValues(fieldInfo.name);
+                    break;
+                case BYTE:
+                    iterator = derivedSourceReaders.getKnnVectorsReader().getByteVectorValues(fieldInfo.name);
+                    break;
+            }
+        } else if (fieldInfo.getDocValuesType() != DocValuesType.NONE && derivedSourceReaders.getDocValuesProducer() != null) { // the field
```

Comment:
why we would have 3 types of field indexes? when this is just vector field.

Reply:
This is not necessarily the vector field. This is for a field in one child before adding back the vector. We are trying to figure out which docId that child maps to. To do it, I get the first field on/after the offset that contains that field. 

So, this is basically field exists - https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/FieldExistsQuery.java